### PR TITLE
Disable Elasticsearch features based on license type

### PR DESCRIPTION
**Pull request by Suraiya-Hameed**
## Description

Disable features based on elastic search license type.

Kibana OIDC and ML jobs are not supported in ES basic license. Add watches to elastic-licensing configmap and disable these features if license type is basic.

## For PR author

- [x] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

File: pkg/controller/utils/utils.go
Comment (Brian-McM) [inline pkg/controller/utils/utils.go:310]: Does the ECK operator create a basic license config map if there's no applied config map? If it does, then you should return "Unknown" here, if it doesn't, then ignore this comment.

**Comment by Brian-McM** [comment]
I'm not 100% sure, but doesn't the intrusion detection job create some required indices and dashboards? It may be that the controller creates certain indices, but https://github.com/tigera/intrusion-detection/blob/master/install/config.yaml#L51 is used for the dashboards.

File: pkg/controller/utils/utils.go
Comment (Suraiya-Hameed) [reply pkg/controller/utils/utils.go:310]: For fresh install of caliEnt, elasticsearch license will be basic by default, plus during the fresh install we won't have any existing `elastic-licensing` configmap until ECK operator is running, so we return "basic" if config map is not found.
Once ECK operator is running, it always creates this config map irrespective of license type. Only possible scenario where config map is not found is during fresh install.

**Comment by Suraiya-Hameed** [comment]
> I'm not 100% sure, but doesn't the intrusion detection job create some required indices and dashboards? It may be that the controller creates certain indices, but https://github.com/tigera/intrusion-detection/blob/master/install/config.yaml#L51 is used for the dashboards.

Thanks for pointing this :+1: 
I've made changes to pass the Elasticsearch license type as env variable to intrusion detection job, there if the license type is basic it will just apply the kibana dashboard related stuff else proceed as usual.

File: pkg/controller/utils/utils.go
Comment (Brian-McM) [reply pkg/controller/utils/utils.go:310]: > Once ECK operator is running, it always creates this config map irrespective of license type

With this in mind I think we should not proceed on anything that requires having knowledge of the license, let's wait until the ECK operator creates the license.

When we have the option to we should avoid assuming assuming configuration for third party software, in this case we just need to wait for ECK to tell us what configuration it will apply. If it changes in the future we won't have a scenario where we temporarily use a basic license then immediately switch to a different one.

File: pkg/controller/utils/utils.go
Comment (Brian-McM) [inline pkg/controller/utils/utils.go:310]: You don't need / shouldn't have a check here for is not found, you need to return an error so that we don't progress, so just return the error as is

**Review by Brian-McM — CHANGES_REQUESTED** [review]

File: pkg/controller/utils/utils.go
Comment (Suraiya-Hameed) [reply pkg/controller/utils/utils.go:310]: This function is called in two controllers and we need this check to differentiate 2 scenario.
- in intrusion-detection, if elastic-licensing is not found or if there is an from this function, we don't install ml jobs
- in logstorage, if elastic-licensing is not found we need to proceed with rendering ECK operator related resource, if there is any other error in `GetElasticLicenseType` we don't progress

Comment (Brian-McM) [reply pkg/controller/utils/utils.go:310]: I think you should just check if the error is a not found one in the operator and continue, and just leave this simple. As far as this function is concerned we should consider not found an error, and if the caller would like to ignore that error then it can do so but checking if it's a not found error.

The code you have of `err != nil || elasticLicenseType == render.ElasticLicenseTypeUnknown` should really just be `err != nil && errors.IsNotFound(err)` (always return Unkown if there's an error).

File: pkg/controller/utils/utils.go
Comment (Brian-McM) [inline pkg/controller/utils/utils.go:321]: I think that you should write a function `render.StrToElasticLicenseType` to guarantee this only returns types that we want to consider. For instance, if don't want to consider somebody who put in a platinum license then this would return unknown if the type was platinum, then we just check if the type is unknown. 

This gives the Unknown type a real meaning besides just what to return for an error, it's Unknown to us but not necessarily an error, but a caller _could_ treat it as an error if it requires a known license type.

There's almost no point in using a specific type if you're just going to cast every string value as the ElasticLicenseType since it doesn't validate that it's a type you've defined.

**Review by Brian-McM — CHANGES_REQUESTED** [review]

**Review by Brian-McM — APPROVED** [review]

File: pkg/render/logstorage.go
Comment (tmjd) [inline pkg/render/logstorage.go:344]: To be clear about what would happen here if elasticLicenseType was Unknown and then went to TypeBasic (and dex was configured).
While the type was unknown we would be configuring dex integration. Would ES just ignore the config or it would error with the config provided?
Once the type was available and Basic the configuration would be updated and ES would run normal.

File: pkg/render/intrusion_detection.go
Comment (tmjd) [inline pkg/render/intrusion_detection.go:200]: Does it matter how Intrusion treats an empty ELASTIC_LICENSE_TYPE?
Is there any problem if initially this uses the wrong License?
It seems likely that this can be ran with the wrong value since it is a job. Since the controller will launch this if the license does not exist, I'm concerned on the impact if it is wrong.
What I'm kind of getting at is, should the controller require the license config?

File: pkg/render/logstorage.go
Comment (Suraiya-Hameed) [reply pkg/render/logstorage.go:344]: > While the type was unknown we would be configuring dex integration. Would ES just ignore the config or it would error with the config provided?

Ideally it should never be unknown as we are reading license values written by ECK operator, if ECK adds support to new license type we will have to catch that during ECK version update and change this condition. Among the licenses that ECK now supports, basic is only type that doesn't support SSO.
If license type is unknown and dex is configured, and say if ES doesn't support OIDC then, ES will work but it will log errors about OIDC and also Kibana will throw license related error.

File: pkg/render/intrusion_detection.go
Comment (tmjd) [reply pkg/render/intrusion_detection.go:200]: Nevermind, I'm wrong. I see the controller bails if the license type is not found.

File: pkg/render/logstorage.go
Comment (tmjd) [reply pkg/render/logstorage.go:344]: Would it be better to ensureLicenseType is Enterprise or EnterpriseTrial? Then we'll be sure we have read the value and won't ever configure Dex without it being a type that supports OIDC?
Is there a downside to do that?

File: pkg/render/intrusion_detection.go
Comment (Suraiya-Hameed) [reply pkg/render/intrusion_detection.go:200]: If ELASTIC_LICENSE_TYPE is set to basic, we skip installing ml jobs in es-job-installer. If it is empty or has a unknown value, it will continue as usual, that is install ml jobs and apply kibana dashboards. So it is basically the same behaviour as before, except we don't install ml jobs if license is basic.

>  Is there any problem if initially this uses the wrong License?

Say if we switch from basic to enterprise, since we are watching `elastic-licensing` configmap, this job will get restarted and ml jobs will get installed, shouldn't be any problem.
For other way around, enterprise to basic, as xpack is not available anymore ml jobs will not run.

I haven't changed anything in the intrusion-detection-controller, it will continue watching for ml jobs and create reports/alerts if job is available, else log error. It does seem to be flood with error, I'll add logic in operator to set `DISABLE_ANOMALY=yes` if license type is basic to avoid all the errors.

File: pkg/render/logstorage.go
Comment (Suraiya-Hameed) [reply pkg/render/logstorage.go]: yes, I've updated to check for Enterprise or EnterpriseTrial.

File: pkg/render/intrusion_detection.go
Comment (tmjd) [inline pkg/render/intrusion_detection.go:374]: Since I don't know what this does can you get someone to review that understands the ML job?

File: pkg/controller/logstorage/logstorage_controller.go
Comment (Brian-McM) [inline pkg/controller/logstorage/logstorage_controller.go:405]: I'm just commenting for the sake of it, but maybe it would make sense for separate controllers for the ECK operator and Elasticsearch?

Please don't consider this for this PR, I'm just putting this thought out inter the ether.

**Review by Brian-McM — APPROVED** [review]

File: pkg/render/intrusion_detection.go
Comment (Brian-McM) [reply pkg/render/intrusion_detection.go:374]: I'm comfortable with this change since watchers aren't supported, but do we also need to set DISABLE_ALERTS to true as well?

**Review by Brian-McM — CHANGES_REQUESTED** [review]
Could you please check if we need to set DISABLE_ALERTS to yes for intrusion detection as well, I think the alerts has a watcher which means with a basic license it will spit out errors.

**Review by tmjd — APPROVED** [review]
LGTM
Please make sure you squash before or when you merge.
One minor thing the Copyright in utils_test.go, should be updated to 2021.

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and leave the branch after I've merged it.

**Comment by marvin-tigera** [comment]
Removing "merge-when-ready" label due to new commits

**Review by Brian-McM — APPROVED** [review]

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and leave the branch after I've merged it.

**Pull request merged**
